job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,217 @@
+import scala.sys.process._
+
+/**
+ * Created by ankits on 4/7/15.
+ */
+class AkkaClusterSupervisorActor(daoActor: ActorRef) extends InstrumentedActor {
+  import ContextSupervisor._
+  import scala.collection.JavaConverters._
+  import scala.concurrent.duration._
+
+  val config = context.system.settings.config
+  val defaultContextConfig = config.getConfig("spark.context-settings")
+  val contextInitTimeout = config.getDuration("spark.context-settings.context-init-timeout",
+                                                TimeUnit.SECONDS)
+  val managerStartCommand = config.getString("deploy.manager-start-cmd")
+  import context.dispatcher
+
+  //actor name -> (context name, context config, context isadhoc)
+  private val contextInitInfos = mutable.HashMap.empty[String, (String, Config, Boolean)]
+  //actor name -> (success callback, failure callback)
+  private val contextCallbacks = mutable.HashMap.empty[String, (ActorRef => Unit, Throwable => Unit)]
+
+  private val contexts = mutable.HashMap.empty[String, ActorRef]
+  private val resultActors = mutable.HashMap.empty[String, ActorRef]
+
+  private val cluster = Cluster(context.system)
+  private val selfAddress = cluster.selfAddress
+
+  // This is for capturing results for ad-hoc jobs. Otherwise when ad-hoc job dies, resultActor also dies,
+  // and there is no way to retrieve results.
+  val globalResultActor = context.actorOf(Props[JobResultActor], "global-result-actor")
+
+  logger.info("AkkaClusterSupervisor initialized on {}", selfAddress)
+
+  override def preStart(): Unit = {
+    cluster.join(selfAddress)
+    cluster.subscribe(self, initialStateMode = InitialStateAsEvents, classOf[MemberEvent])
+  }
+
+  override def postStop(): Unit = {
+    cluster.unsubscribe(self)
+  }
+
+  def wrappedReceive: Receive = {
+    case MemberUp(member) =>
+      if (member.hasRole("manager")) {
+        val memberActors = RootActorPath(member.address) / "user" / "*"
+        context.actorSelection(memberActors) ! Identify(memberActors)
+      }
+
+    case ActorIdentity(memberActors, actorRefOpt) =>
+      actorRefOpt.map { actorRef =>
+        val actorName = actorRef.path.name
+        if (actorName.startsWith("jobManager")) {
+          logger.info("Received identify response, attempting to initialize context at {}", memberActors)
+          val initInfoOpt = contextInitInfos.remove(actorName)
+          val callbackOpt = contextCallbacks.remove(actorName)
+          if (initInfoOpt.isDefined && callbackOpt.isDefined) {
+            val (ctxName, ctxConf, isAdHoc) = initInfoOpt.get
+            val (successFunc, failureFunc) = callbackOpt.get
+            initContext(actorName, actorRef, contextInitTimeout)(
+              ctxName, ctxConf, isAdHoc)(successFunc, failureFunc)
+          }
+          else {
+            logger.warn("No initialization or callback found for jobManager actor {}", actorRef.path)
+            actorRef ! PoisonPill
+          }
```

Comment:
Hrm, you do something similar above. I'd recommend a for comp. like this vs if/else, and lose the '.get' calls, unless you can insure they will never result in a None and throw an error.
      
     (for (a <- opt1; b <- opt2) yield initContext(...)) getOrElse {
         logger.warn("No initialization or callback found for jobManager actor {}", actorRef.path)
         actorRef ! PoisonPill
      }